<a href="https://colab.research.google.com/github/Ammarakanwal/FlyRank_Repo/blob/main/Copy_of_w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Ammarakanwal/FlyRank_Repo/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [ ]:
from google.colab import files

uploaded = files.upload()

Saving content_refresh_anonymized.csv to content_refresh_anonymized (1).csv


In [ ]:
import pandas as pd

df = pd.read_csv("content_refresh_anonymized.csv")
df.head()




,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [ ]:
df.shape

(30000, 44)

In [ ]:
df.describe()

,search_volume,competition,cpc,word_count,char_count,impressions_90d,clicks_90d,pageviews_90d,sessions_90d,users_90d,...,sessions_prev_30d,content_age_days,age_tier_order,days_since_last_update,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,trend_pct
count,27532.000000,27532.000000,27532.000000,22301.000000,22301.000000,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,...,30000.000000,30000.00000,30000.000000,30000.000000,30000.000000,30000.00000,30000.000000,29875.000000,30000.000000,26612.000000
mean,158.882391,0.146954,0.485342,3107.760325,20665.277835,5200.366300,16.097333,49.942467,37.066633,35.937700,...,10.283000,256.16780,4.786533,46.098300,0.510733,16.34238,2.534520,18.212921,0.768196,-4.785969
std,1518.270825,0.285241,2.101560,1452.382598,10115.344042,16838.019547,75.076958,152.101430,107.069131,103.748185,...,42.578003,132.70793,0.790392,42.078709,3.279162,15.21679,8.310096,29.472768,7.429454,473.861780
min,0.000000,0.000000,0.000000,8.000000,40.000000,1.000000,0.000000,0.000000,1.000000,1.000000,...,0.000000,90.00000,3.000000,1.000000,0.000000,0.00000,0.000000,0.000000,0.000000,-100.000000
25%,0.000000,0.000000,0.000000,2413.000000,15644.000000,81.000000,0.000000,2.000000,2.000000,2.000000,...,1.000000,132.00000,4.000000,20.000000,0.000000,6.20000,0.000000,0.000000,0.000000,-62.600000
50%,10.000000,0.000000,0.000000,2877.000000,19116.000000,731.000000,1.000000,8.000000,7.000000,7.000000,...,2.000000,236.00000,5.000000,20.000000,0.070000,10.80000,0.000000,5.000000,0.000000,-33.500000
75%,20.000000,0.130000,0.000000,3666.000000,24011.000000,3615.250000,7.000000,33.000000,27.000000,27.000000,...,7.000000,333.00000,5.000000,104.000000,0.290000,22.30000,1.350000,23.530000,0.000000,0.000000
max,74000.000000,1.000000,100.360000,9546.000000,111158.000000,517715.000000,4178.000000,5998.000000,4345.000000,4913.000000,...,4247.000000,564.00000,6.000000,373.000000,100.000000,245.00000,100.000000,300.000000,300.000000,44900.000000


In [ ]:
df["impressions_90d"].describe()

,impressions_90d
count,30000.000000
mean,5200.366300
std,16838.019547
min,1.000000
25%,81.000000
50%,731.000000
75%,3615.250000
max,517715.000000


In [ ]:
df["trend_direction"].value_counts()

,count
trend_direction,
down,16262
stable,5962
up,4388
new,2236
flat,1152


# My Lane

I selected **Refresh / Content Opportunity Scoring** as my project lane.

The objective of this project is to identify webpages that should be reviewed first for content refresh or optimization. Since organizations often have thousands of webpages but limited time and resources, prioritizing the right pages can help improve search visibility more efficiently.

I selected this lane because it combines business decision-making with data analysis and machine learning to support evidence-based SEO recommendations.

# Research Question

## Search Question

Can historical SEO and engagement metrics be used to identify webpages that should be prioritized for content refresh?

## Decision

Which webpages should an SEO team review first?

## Unit of Analysis

One webpage (one content page).

## Output

A ranked list of webpages according to their refresh priority.

## Action

The SEO team can refresh, improve, expand, monitor or optimize the highest-ranked pages first.

## Cost of a Wrong Recommendation

If an important page is not selected, valuable traffic and business opportunities may be lost.

If an unimportant page is selected, time and resources will be wasted.

## Why Data or Machine Learning?

The dataset contains many measurable SEO indicators including impressions, clicks, CTR, average position, engagement rate and content age. Instead of manually reviewing every webpage, machine learning can recognize useful patterns and help prioritize pages more efficiently.

# Quick Look at the Data

The starter dataset contains **30,000 webpages** with **44 variables**.

Some initial observations include:

- The average number of impressions over the last 90 days is **5,200.37**.
- The median number of impressions is **731**.
- The maximum number of impressions is **517,715**, showing that webpage visibility varies greatly.
- The dataset contains **16,262 declining pages**, **5,962 stable pages**, **4,388 growing pages**, **2,236 new pages**, and **1,152 flat pages**.

These statistics suggest that the dataset contains substantial variation in webpage performance, making it suitable for investigating content refresh opportunities.

# Careful Words

This project analyzes historical SEO performance data to identify patterns associated with webpages that may require content review.

The results should not be interpreted as proof that refreshing a webpage will automatically increase traffic. Search performance can also be influenced by Google's algorithms, competition, seasonality and user behavior.

Therefore, the recommendations produced by this project are intended to support human decision-making rather than replace it.

# Self-Check

✅ I selected one project lane.

✅ I clearly defined my research question.

✅ I identified the business decision.

✅ I defined the unit of analysis.

✅ I explained the action someone could take.

✅ I described the cost of a wrong recommendation.

✅ I supported my lane choice using real numbers from the dataset.

✅ I avoided making causal claims.

✅ I explained why machine learning may be useful.

In [ ]:
print("Dataset Shape:", df.shape)

print("\nAverage Impressions:")
print(df["impressions_90d"].mean())

print("\nMedian Impressions:")
print(df["impressions_90d"].median())

print("\nTrend Distribution:")
print(df["trend_direction"].value_counts())

Dataset Shape: (30000, 44)

Average Impressions:
5200.3663

Median Impressions:
731.0

Trend Distribution:
trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64
